File: LeetCode/Normal/145.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    vector<int> postorderTraversal(TreeNode *root)
    {
        stack<TreeNode *> s;
        vector<int> nums;
        TreeNode *pre = NULL;
        while (!s.empty() || root)
        {
            if (root)
            {
                s.push(root);
                root = root->left;
            }
            else
            {
                root = s.top();
                if (!root->right || pre == root->right)
                {
                    pre = root;
                    nums.push_back(root->val);
                    s.pop();
                    root = NULL;
                }
                else
                    root = root->right;
            }
        }
        return nums;
    }
};

File: LeetCode/Normal/100-199/145.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        lst = []
        pre = None
        stack = []
        while root or len(stack) > 0:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack[-1]
                # 说明访问过子节点回退
                if root.right==None or root.right == pre:
                    stack.pop()
                    pre = root
                    lst.append(root.val)
                    root=None
                else:
                    root = root.right
        return lst